File: src/main/kotlin/fr/ignishky/framework/cqrs/event/Event.java
package fr.ignishky.framework.cqrs.event;

import fr.ignishky.framework.domain.Aggregate;

import java.time.Instant;

public abstract class Event<I, A extends Aggregate<I>, P extends Payload> {

    protected final String id;
    protected final I aggregateId;
    private final Class<A> aggregateClass;
    protected final P payload;
    protected final Instant instant;

    protected Event(String id, I aggregateId, Class<A> aggregateClass, P payload, Instant instant) {
        this.id = id;
        this.aggregateId = aggregateId;
        this.aggregateClass = aggregateClass;
        this.payload = payload;
        this.instant = instant;
    }

    public abstract A apply(A aggregate);

}

File: src/main/kotlin/fr/ignishky/framework/cqrs/event/Event.kt
package fr.ignishky.framework.cqrs.event

import fr.ignishky.framework.domain.Aggregate
import fr.ignishky.framework.domain.AggregateId
import java.time.Instant
import kotlin.reflect.KClass

abstract class Event<I : AggregateId, A : Aggregate<I>, P : Payload>
protected constructor(
    val id: Long,
    val aggregateId: I,
    val aggregateClass: KClass<A>,
    val payload: P,
    val instant: Instant
) {

    abstract fun apply(aggregate: A): A

    override fun toString(): String {
        return "Event($id, $aggregateId, $aggregateClass, $payload, $instant)"
    }

}
